nl_server/loader.py

```diff
@@ -35,39 +38,61 @@
 #
 # Reads the yaml files and loads all the server state.
 #
-def load_server_state(app: Any):
+def load_server_state(app: Flask):
   flask_env = os.environ.get('FLASK_ENV')
 
   embeddings_map, models_map = _load_yamls(flask_env)
 
   # In local dev, cache the embeddings on disk so each hot reload won't download
   # the embeddings again.
   if _use_cache(flask_env):
-    from diskcache import Cache
     cache = Cache(NL_CACHE_PATH, size_limit=_NL_CACHE_SIZE_LIMIT)
     cache.expire()
 
     nl_model = cache.get(NL_MODEL_CACHE_KEY)
     nl_embeddings = cache.get(NL_EMBEDDINGS_CACHE_KEY)
     if nl_model and nl_embeddings:
-      app.config[config.NL_MODEL_KEY] = nl_model
-      app.config[config.NL_EMBEDDINGS_KEY] = nl_embeddings
-      app.config[config.NL_EMBEDDINGS_VERSION_KEY] = embeddings_map
+      _update_app_config(app, nl_model, nl_embeddings, embeddings_map)
       return
 
   nl_embeddings = embeddings_store.Store(config.load(embeddings_map,
                                                      models_map))
   nl_model = NLAttributeModel()
-  app.config[config.NL_MODEL_KEY] = nl_model
-  app.config[config.NL_EMBEDDINGS_KEY] = nl_embeddings
-  app.config[config.NL_EMBEDDINGS_VERSION_KEY] = embeddings_map
+  _update_app_config(app, nl_model, nl_embeddings, embeddings_map)
+
+  _maybe_cache(flask_env, nl_embeddings, nl_model)
 
-  if _use_cache(flask_env):
-    with Cache(cache.directory, size_limit=_NL_CACHE_SIZE_LIMIT) as reference:
-      reference.set(NL_EMBEDDINGS_CACHE_KEY,
-                    nl_embeddings,
-                    expire=_NL_CACHE_EXPIRE)
-      reference.set(NL_MODEL_CACHE_KEY, nl_model, expire=_NL_CACHE_EXPIRE)
+
+def load_custom_embeddings(app: Flask):
+  """Loads custom DC embeddings at runtime.
+
+  This method should only be called at runtime (i.e. NOT at startup).
+  It assumes that embeddings were already initialized at startup and this method
+  only merges any newly loaded custom embeddings with the default embeddings.
+  """
+  flask_env = os.environ.get('FLASK_ENV')
+  embeddings_map, _ = _load_yamls(flask_env)
+  custom_embeddings_local_path = embeddings_map.get(config.CUSTOM_DC_INDEX)
+  if not custom_embeddings_local_path:
+    logging.warning("No custom DC embeddings found, so none will be loaded.")
+    return
+
+  custom_idx = config.EmbeddingsIndex(
+      name=config.CUSTOM_DC_INDEX,
+      embeddings_file_name=os.path.basename(custom_embeddings_local_path),
+      embeddings_local_path=custom_embeddings_local_path)
```

Comment:
Is this now making an assumption that `custom_embeddings_local_path` always gets updated?  Also that it is always local?

Can we check we're in "local GKE" or "RSI" mode?  Or use `config._parse()`, and it will then decide whether the path is local or not.

Reply:
Chatted offline. Called out in comment and added a TODO to call `config._parse()` instead.